### ISSUE TITLE
Nepali character width calculated incorrectly

### ISSUE THREAD

**Issue by danielbachhuber** [wp-cli/php-cli-tools#103]
![term](https://cloud.githubusercontent.com/assets/2098823/16546177/07fbabca-4162-11e6-8bd1-51c356b1aa2a.png)

From https://github.com/wp-cli/wp-cli/issues/3038#issuecomment-230158804

**Comment by johnbillion** [comment]
Yay! Character encoding!

This is due to combining marks in the string. `mb_strlen()` counts characters, but when a combining mark is printed it gets combined with another character to form a grapheme with the width of one character. That's what we actually want to count in a situation like this where the number of printed characters is important.

`mb_strlen()` will be short by one character for every combining mark present in the string.

You can count the number of graphemes in a string with `grapheme_strlen()` but this requires the `intl` extension, and I've no idea how widespread that is. Stack Overflow tells me that `preg_match_all( '/\X/u', $str)` is an alternative.

The character counting issue in php-cli-tools is in [`\cli\safe_strlen()`](https://github.com/wp-cli/php-cli-tools/blob/28880df59f3c8c64db758f8e93cf5feaf3ddb310/lib/cli/cli.php#L155-L173).

**Comment by johnbillion** [comment]
This might be a solution, but it's untested. I'll take a proper look later:

```
function safe_strlen( $str ) {
	return preg_match_all( '/\X/u', $str );
}
```

**Comment by danielbachhuber** [comment]
> Yay! Character encoding!

I think I detect sarcasm here but I'm not quite sure... ;)

> This might be a solution, but it's untested. I'll take a proper look later

Sounds good, thanks!

**Comment by johnbillion** [comment]
Wow. Turns out that the font makes a difference to how some combined characters appear. Here's the same output in two different fonts:

SF Mono:
<img width="166" alt="screenshot 2017-02-17 17 20 54" src="https://cloud.githubusercontent.com/assets/208434/23075751/90d1a6f2-f535-11e6-83b5-6a816c37cd90.png">

Menlo:
<img width="189" alt="screenshot 2017-02-17 17 21 02" src="https://cloud.githubusercontent.com/assets/208434/23075750/90d17cfe-f535-11e6-96fe-03abdf636b2e.png">

**Comment by danielbachhuber** [comment]
> Turns out that the font makes a difference to how some combined characters appear.

:(

**Comment by gitlost** [comment]
I don't think there's much one can do about fonts not displaying stuff correctly but I got the original Nepali example working (on Ubuntu at least) by using the suggested `grapheme_strlen()` (with `preg_match_all( '/\X/u' )` backup) in a new function `strwidth()`, to be called by `safe_str_pad()`, with adjustments for [East Asian Width](http://www.unicode.org/reports/tr11/tr11-19.html).

![screenshot from 2017-07-22 23-43-54](https://user-images.githubusercontent.com/481982/28495352-ce4ef238-6f41-11e7-8224-a28be52b5c99.png)

PR to follow.

**Comment by schlessera** [comment]
Resolved (as good as possible) through #107 .
